File: lintcode/python/1399_take_coins.py
#!/usr/bin/python -t

# two pointers
# 这题的正解因该是用双指针来维护一个虚拟的滑动窗口。说它是虚拟的是因为这个窗口大部分时间不连续。

# 窗口可以从尾部开始向右移动，也可以从头部开始向左移动。

class Solution:
    """
    @param list: The coins
    @param k: The k
    @return: The answer
    """
    def takeCoins(self, list, k):
        # Write your code here
        n = len(list)
        if n <= k:
            return sum(list)
            
        lo = -k
        hi = -1
        ret = running_coin = sum(list[lo:])
        
        while lo != 0:
            running_coin -= list[lo]
            lo += 1
            hi += 1
            running_coin += list[hi]
            
            ret = max(ret, running_coin)
            
        return ret
        

File: lintcode/golang/1399_take_coins.go
/**
 * @param list: The coins
 * @param k: The k
 * @return: The answer
 */
func takeCoins (list []int, k int) int {
    // Write your code here
    n := len(list)
    if n <= k {
        return sum(list)
    }
    
    l := -1
    r := -k
    
    all_right := sum(list[n+r:])
    ret := all_right
    
    for r != 0 {
        all_right -= list[n+r]
        r++
        l++
        all_right += list[l]
        ret = max(ret, all_right)
    }
    
    return ret
}

func sum(l []int) int {
    ret := 0
    
    for _, v := range l {
        ret += v
    }
    
    return ret
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}
